File: agent/local_config.py
from __future__ import annotations

import json
import time
from contextlib import suppress
from dataclasses import dataclass
from pathlib import Path
from typing import Any
from uuid import uuid4

LOCAL_CONFIG_FILE = "emperor.local.json"


@dataclass(frozen=True)
class WebUIPreferences:
    host: str = "127.0.0.1"
    port: int = 8765
    open_browser: bool = False


@dataclass(frozen=True)
class DesktopPetPreferences:
    enabled: bool = False
    auto_start_with_webui: bool = True


@dataclass(frozen=True)
class LocalConfig:
    webui: WebUIPreferences = WebUIPreferences()
    desktop_pet: DesktopPetPreferences = DesktopPetPreferences()


def load_local_config(root: Path) -> LocalConfig:
    path = Path(root).resolve() / LOCAL_CONFIG_FILE
    if not path.exists():
        return LocalConfig()
    try:
        raw = json.loads(path.read_text(encoding="utf-8") or "{}")
    except (json.JSONDecodeError, OSError):
        _preserve_corrupt_local_config(path)
        return LocalConfig()
    return parse_local_config(raw)


def parse_local_config(raw: dict[str, Any] | None) -> LocalConfig:
    data = raw if isinstance(raw, dict) else {}
    webui = data.get("webui") if isinstance(data.get("webui"), dict) else {}
    desktop_pet = data.get("desktopPet")
    if not isinstance(desktop_pet, dict):
        desktop_pet = data.get("desktop_pet")
    if not isinstance(desktop_pet, dict):
        desktop_pet = {}
    return LocalConfig(
        webui=WebUIPreferences(
            host=str(webui.get("host") or "127.0.0.1"),
            port=_valid_port(webui.get("port"), 8765),
            open_browser=bool(webui.get("openBrowser", webui.get("open_browser", False))),
        ),
        desktop_pet=DesktopPetPreferences(
            enabled=bool(desktop_pet.get("enabled", False)),
            auto_start_with_webui=bool(
                desktop_pet.get(
                    "autoStartWithWebui",
                    desktop_pet.get("auto_start_with_webui", True),
                )
            ),
        ),
    )


def save_local_config(root: Path, config: LocalConfig) -> Path:
    path = Path(root).resolve() / LOCAL_CONFIG_FILE
    payload = {
        "webui": {
            "host": config.webui.host,
            "port": config.webui.port,
            "openBrowser": config.webui.open_browser,
        },
        "desktopPet": {
            "enabled": config.desktop_pet.enabled,
            "autoStartWithWebui": config.desktop_pet.auto_start_with_webui,
        },
    }
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_name(f".{path.name}.{uuid4().hex}.tmp")
    tmp.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    tmp.replace(path)
    return path


def merge_webui_overrides(
    config: LocalConfig,
    *,
    host: str | None = None,
    port: int | None = None,
    open_browser: bool | None = None,
) -> WebUIPreferences:
    return WebUIPreferences(
        host=str(host or config.webui.host or "127.0.0.1"),
        port=_valid_port(port if port is not None else config.webui.port, 8765),
        open_browser=config.webui.open_browser if open_browser is None else bool(open_browser),
    )


def local_config_path(root: Path) -> Path:
    return Path(root).resolve() / LOCAL_CONFIG_FILE


def local_config_diagnostics(root: Path) -> dict[str, Any]:
    path = local_config_path(root)
    backups = sorted(
        path.parent.glob(f"{LOCAL_CONFIG_FILE}.corrupt-*"),
        key=lambda item: item.stat().st_mtime,
        reverse=True,
    )
    status = "missing"
    error = ""
    if path.exists():
        try:
            json.loads(path.read_text(encoding="utf-8") or "{}")
        except (json.JSONDecodeError, OSError) as exc:
            status = "corrupt"
            error = str(exc)
        else:
            status = "ok"
    return {
        "path": path.as_posix(),
        "exists": path.exists(),
        "status": status,
        "error": error,
        "corruptBackups": [
            {
                "path": item.as_posix(),
                "bytes": item.stat().st_size,
                "updatedAt": item.stat().st_mtime,
            }
            for item in backups[:10]
        ],
    }


def _preserve_corrupt_local_config(path: Path) -> None:
    backup = path.with_name(f"{path.name}.corrupt-{int(time.time())}-{uuid4().hex[:8]}")
    with suppress(OSError):
        if path.exists():
            path.rename(backup)


def _valid_port(value: Any, default: int) -> int:
    try:
        port = int(value)
    except (TypeError, ValueError):
        return default
    if 1 <= port <= 65535:
        return port
    return default

File: packages/core/src/config/local-config.ts
import { randomUUID } from 'node:crypto'
import { existsSync } from 'node:fs'
import {
  mkdir,
  readFile,
  readdir,
  rename,
  stat,
  writeFile,
} from 'node:fs/promises'
import { dirname, join, resolve } from 'node:path'
import {
  parsePermissionRules,
  type PermissionRuleDiagnostics,
  type PermissionRuleInput,
} from '../permissions/rules'

export const LOCAL_CONFIG_FILE = 'emperor.local.json'

export interface WebUIPreferences {
  host: string
  port: number
  openBrowser: boolean
}

export interface DesktopPetPreferences {
  enabled: boolean
  autoStartWithWebui: boolean
}

export type PromptProfile = 'classic' | 'neutral' | 'technical'

export interface PromptPreferences {
  profile: PromptProfile
}

export interface PermissionPreferences {
  rules: PermissionRuleInput[]
}

export interface LocalConfig {
  webui: WebUIPreferences
  desktopPet: DesktopPetPreferences
  prompt: PromptPreferences
  permissions: PermissionPreferences
}

export interface LocalConfigBackup {
  path: string
  bytes: number
  updatedAt: number
}

export interface LocalConfigDiagnostics {
  path: string
  exists: boolean
  status: 'missing' | 'ok' | 'corrupt'
  error: string
  permissions: PermissionRuleDiagnostics
  corruptBackups: LocalConfigBackup[]
}

function defaultLocalConfig(): LocalConfig {
  return {
    webui: { host: '127.0.0.1', port: 8765, openBrowser: false },
    desktopPet: { enabled: false, autoStartWithWebui: true },
    prompt: { profile: 'technical' },
    permissions: { rules: [] },
  }
}

function objectOrEmpty(value: unknown): Record<string, any> {
  return value && typeof value === 'object' && !Array.isArray(value)
    ? (value as Record<string, any>)
    : {}
}

function validPort(value: unknown, fallback: number): number {
  const port =
    typeof value === 'number'
      ? Math.trunc(value)
      : Number.parseInt(String(value), 10)
  return Number.isFinite(port) && port >= 1 && port <= 65535 ? port : fallback
}

export function parseLocalConfig(
  raw: Record<string, any> | null | undefined,
): LocalConfig {
  const data = objectOrEmpty(raw)
  const webui = objectOrEmpty(data.webui)
  let desktopPet = objectOrEmpty(data.desktopPet)
  if (Object.keys(desktopPet).length === 0)
    desktopPet = objectOrEmpty(data.desktop_pet)
  const prompt = objectOrEmpty(data.prompt)
  const permissions = objectOrEmpty(data.permissions)
  return {
    webui: {
      host: String(webui.host || '127.0.0.1'),
      port: validPort(webui.port, 8765),
      openBrowser: Boolean(webui.openBrowser ?? webui.open_browser ?? false),
    },
    desktopPet: {
      enabled: Boolean(desktopPet.enabled ?? false),
      autoStartWithWebui: Boolean(
        desktopPet.autoStartWithWebui ??
        desktopPet.auto_start_with_webui ??
        true,
      ),
    },
    prompt: {
      profile: normalizePromptProfile(prompt.profile),
    },
    permissions: {
      rules: Array.isArray(permissions.rules)
        ? (permissions.rules.filter(
            (item) => item && typeof item === 'object' && !Array.isArray(item),
          ) as PermissionRuleInput[])
        : [],
    },
  }
}

export function localConfigPath(root: string): string {
  return join(resolve(root), LOCAL_CONFIG_FILE)
}

export async function loadLocalConfig(
  root: string,
  opts: { preserveCorrupt?: boolean } = {},
): Promise<LocalConfig> {
  const path = localConfigPath(root)
  if (!existsSync(path)) return defaultLocalConfig()
  try {
    return parseLocalConfig(JSON.parse((await readFile(path, 'utf8')) || '{}'))
  } catch {
    if (opts.preserveCorrupt !== false) await preserveCorruptLocalConfig(path)
    return defaultLocalConfig()
  }
}

export async function saveLocalConfig(
  root: string,
  config: LocalConfig,
): Promise<string> {
  const path = localConfigPath(root)
  const payload = {
    webui: {
      host: config.webui.host,
      port: config.webui.port,
      openBrowser: config.webui.openBrowser,
    },
    desktopPet: {
      enabled: config.desktopPet.enabled,
      autoStartWithWebui: config.desktopPet.autoStartWithWebui,
    },
    prompt: {
      profile: normalizePromptProfile(config.prompt?.profile),
    },
    permissions: {
      rules: Array.isArray(config.permissions?.rules)
        ? config.permissions.rules
        : [],
    },
  }
  await mkdir(dirname(path), { recursive: true })
  const tmp = join(
    dirname(path),
    `.${LOCAL_CONFIG_FILE}.${randomUUID().replace(/-/g, '')}.tmp`,
  )
  await writeFile(tmp, `${JSON.stringify(payload, null, 2)}\n`, 'utf8')
  await rename(tmp, path)
  return path
}

export function normalizePromptProfile(value: unknown): PromptProfile {
  return value === 'classic' || value === 'neutral' || value === 'technical'
    ? value
    : 'technical'
}

export function mergeWebuiOverrides(
  config: LocalConfig,
  overrides: {
    host?: string | null
    port?: number | null
    openBrowser?: boolean | null
  } = {},
): WebUIPreferences {
  return {
    host: String(overrides.host || config.webui.host || '127.0.0.1'),
    port: validPort(overrides.port ?? config.webui.port, 8765),
    openBrowser:
      overrides.openBrowser === null || overrides.openBrowser === undefined
        ? config.webui.openBrowser
        : Boolean(overrides.openBrowser),
  }
}

export async function localConfigDiagnostics(
  root: string,
): Promise<LocalConfigDiagnostics> {
  const path = localConfigPath(root)
  const exists = existsSync(path)
  let status: LocalConfigDiagnostics['status'] = 'missing'
  let error = ''
  if (exists) {
    try {
      const raw = JSON.parse((await readFile(path, 'utf8')) || '{}')
      const parsed = parseLocalConfig(raw)
      const permissionDiagnostics = parsePermissionRules(
        parsed.permissions.rules,
      ).diagnostics
      status = 'ok'
      return {
        path,
        exists,
        status,
        error,
        permissions: permissionDiagnostics,
        corruptBackups: await listCorruptBackups(path),
      }
    } catch (err) {
      status = 'corrupt'
      error = err instanceof Error ? err.message : String(err)
    }
  }
  return {
    path,
    exists,
    status,
    error,
    permissions: parsePermissionRules([]).diagnostics,
    corruptBackups: await listCorruptBackups(path),
  }
}

async function preserveCorruptLocalConfig(path: string): Promise<void> {
  if (!existsSync(path)) return
  const seconds = Math.trunc(Date.now() / 1000)
  const suffix = randomUUID().replace(/-/g, '').slice(0, 8)
  await rename(path, `${path}.corrupt-${seconds}-${suffix}`).catch(() => {})
}

async function listCorruptBackups(path: string): Promise<LocalConfigBackup[]> {
  const parent = dirname(path)
  const prefix = `${LOCAL_CONFIG_FILE}.corrupt-`
  const names = await readdir(parent).catch(() => [])
  const backups: LocalConfigBackup[] = []
  for (const name of names) {
    if (!name.startsWith(prefix)) continue
    const fullPath = join(parent, name)
    const info = await stat(fullPath).catch(() => null)
    if (!info) continue
    backups.push({
      path: fullPath,
      bytes: info.size,
      updatedAt: info.mtimeMs / 1000,
    })
  }
  backups.sort((a, b) => b.updatedAt - a.updatedAt)
  return backups.slice(0, 10)
}
